ir/src/boundary_stmts.rs

```diff
@@ -0,0 +1,175 @@
+use crate::TraceSegment;
+
+use super::{BTreeMap, BoundaryExpr, IdentifierType, SemanticError, SymbolTable};
+use parser::ast::{self, BoundaryStmt};
+
+// BOUNDARY CONSTRAINTS
+// ================================================================================================
+
+/// A struct containing all of the boundary constraints to be applied at each of the 2 allowed
+/// boundaries (first row and last row). The constraints are stored in a vector of tuples where
+/// each index of the vector corresponds to the trace segment of the column the boundary constraint
+/// applies to. For ease of code generation and evaluation, constraints are sorted into maps by the
+/// boundary. This also simplifies ensuring that there are no conflicting constraints sharing a
+/// boundary and column index.
+#[derive(Default, Debug)]
+pub(crate) struct BoundaryStmts {
+    boundary_constraints: Vec<(BTreeMap<usize, BoundaryExpr>, BTreeMap<usize, BoundaryExpr>)>,
+}
+
+impl BoundaryStmts {
+    /// Creates a new [BoundaryStmts] instance with the specified number of trace segments.
+    pub fn new(num_trace_segments: usize) -> Self {
+        Self {
+            boundary_constraints: vec![(BTreeMap::new(), BTreeMap::new()); num_trace_segments],
+        }
+    }
+
+    // --- ACCESSORS ------------------------------------------------------------------------------
+
+    /// Returns the number of boundary constraints for the specified trace segment.
+    pub fn num_boundary_constraints(&self, trace_segment: TraceSegment) -> usize {
+        if self.boundary_constraints.len() <= trace_segment.into() {
+            0
+        } else {
+            self.boundary_constraints[trace_segment as usize].0.len()
+                + self.boundary_constraints[trace_segment as usize].1.len()
```

Comment:
Maybe it would be clearer to use a struct instead of a tuple so we could refer to first and last instead of 0 and 1. Wdyt?